source/Nuke.Common/Git/GitRepository.cs

```diff
@@ -51,12 +45,53 @@ public static GitRepository FromLocalDirectory(string directory, string branch =
 
             var (endpoint, identifier) = ParseUrl(url);
 
+            var parsedHead = ParseHeadFile(gitDirectory, branch, commit);
+
             return new GitRepository(
                 endpoint,
                 identifier,
                 rootDirectory,
-                head,
-                branch ?? (branchMatch.Success ? branchMatch.Groups["branch"].Value : null));
+                parsedHead.Head,
+                parsedHead.Branch,
+                parsedHead.Commit);
+        }
+
+        private static (string Head, string Branch, string Commit) ParseHeadFile(string gitDirectory, string branch, string commit)
+        {
+            var headFile = Path.Combine(gitDirectory, "HEAD");
+            ControlFlow.Assert(File.Exists(headFile), $"File.Exists({headFile})");
+            var head = File.ReadAllLines(headFile).First();
+
+            var attachedHead = Regex.Match(head, "^ref: (?<reference>refs/heads/(?<branch>.*))");
+
+            // Don't parse branch if it has been supplied from outer scope
+            if (branch == null)
+            {
+                if (attachedHead.Success)
+                {
+                    branch = attachedHead.Groups["branch"].Value;
+                }
+            }
+
+            // Don't parse commit if it has been supplied from outer scope
+            if (commit == null)
+            {
+                // Attached, follow the ref for the commit
+                if (attachedHead.Success)
+                {
+                    var headRefFile = Path.Combine(gitDirectory, attachedHead.Groups["reference"].Value);
+                    ControlFlow.Assert(File.Exists(headRefFile), $"File.Exists({headRefFile})");
+
+                    commit = File.ReadAllLines(headFile).First();
```

Comment:
`headRefFile`?

Reply:
Couldn't figure out a better name..

It is the file being referenced by HEAD file

(.git/HEAD)
ref: refs/heads/gitrepo-commit

(.git\refs\heads\gitrepo-commit)
0460fa91473c6334d012caf81b69c42bd35103ec

Reply:
That’s not what I mean. Currently Head and Commit both read from the same file?

Reply:
Indeed. And they shouldn't - Fixed and pushed